[rb] Add URLs constant to update error messages

aguspe opened the pull request:
### **User description**
### Description
This PR adds a URLs constant on the error module to be able to obtain the URL based on a symbol for each WebDriver error subclass

### Motivation and Context
Based on #11512 the goal is to provide more information in the error message by linking to the selenium documentation

### Types of changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

### Checklist
- [x] I have read the [contributing](https://github.com/SeleniumHQ/selenium/blob/trunk/CONTRIBUTING.md) document.
- [x] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.


___

### **PR Type**
Enhancement, Tests


___

### **Description**
- Added a `url` method to the `WebDriverError` class to generate URLs dynamically based on the class name.
- Modified error messages to include the dynamically generated URLs.
- Removed hardcoded URLs from specific error classes such as `NoSuchElementError`, `StaleElementReferenceError`, `InvalidSelectorError`, and `NoSuchDriverError`.
- Added tests to verify the correct URL generation for error messages.
- Updated type signatures to reflect the new methods in the `WebDriverError` class.



___



### **Changes walkthrough** 📝
<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Enhancement
</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>error.rb</strong><dd><code>Add dynamic URL generation for error messages.</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>
      
rb/lib/selenium/webdriver/common/error.rb

<li>Added <code>url</code> method to <code>WebDriverError</code> class to generate URLs based on <br>class names.<br> <li> Modified error messages to include generated URLs.<br> <li> Removed hardcoded URLs from specific error classes.<br>


</details>
    

  </td>
  <td><a href="https://github.com/SeleniumHQ/selenium/pull/14174/files#diff-942df34e1c3e79a3a6ada80fc46e4ef30a799c7a133efdff252f8b3cce4d7ea1">+22/-21</a>&nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>error.rbs</strong><dd><code>Update type signatures for new methods in `WebDriverError`.</code></dd></summary>
<hr>
      
rb/sig/lib/selenium/webdriver/common/error.rbs

<li>Updated type signatures to include new methods in <code>WebDriverError</code>.<br>


</details>
    

  </td>
  <td><a href="https://github.com/SeleniumHQ/selenium/pull/14174/files#diff-d1b5c312e16beda54108fbca95587dd70069a93e6efadde7fe234b13c13ed976">+5/-0</a>&nbsp; &nbsp; &nbsp; </td>

</tr>                    
</table></td></tr><tr><td><strong>Tests
</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>error_spec.rb</strong><dd><code>Add tests for dynamic URL generation in error messages.</code>&nbsp; &nbsp; </dd></summary>
<hr>
      
rb/spec/integration/selenium/webdriver/error_spec.rb

<li>Added tests for self-generated URLs in error messages.<br> <li> Included test for <code>NoSuchElementError</code> to verify URL generation.<br>


</details>
    

  </td>
  <td><a href="https://github.com/SeleniumHQ/selenium/pull/14174/files#diff-9a507396bce1fb2360f6852a8022c9df9dcaf6869be5a95feef6571adb2bcd20">+11/-0</a>&nbsp; &nbsp; </td>

</tr>                    
</table></td></tr></tr></tbody></table>

___

> 💡 **PR-Agent usage**:
>Comment `/help` on the PR to get a list of all available PR-Agent tools and their descriptions

p0deje reviewed (commented):
I see that you dynamically generate URLs from a class name of the error. I have a feeling it might be fragile and prone to subtle bugs if somebody makes a typo in a URL while working on #1276. It also make it really hard to refactor the website and make sure URLs are used correctly in all bindings. The problem is that this can't be grepped anymore.

I'd suggest we instead have a hash of error => URL mapping that will explicitly allow us to mark which errors already have URLs and add them as we go.

```ruby
URLS = {
  NoSuchElementError => '#no-such-element-exception',
  ...
}
```

Looking forward to what @titusfortner thinks on that?

aguspe commented:
> I see that you dynamically generate URLs from a class name of the error. I have a feeling it might be fragile and prone to subtle bugs if somebody makes a typo in a URL while working on #1276. It also make it really hard to refactor the website and make sure URLs are used correctly in all bindings. The problem is that this can't be grepped anymore.
> 
> 
> 
> I'd suggest we instead have a hash of error => URL mapping that will explicitly allow us to mark which errors already have URLs and add them as we go.
> 
> 
> 
> ```ruby
> 
> URLS = {
> 
>   NoSuchElementError => '#no-such-element-exception',
> 
>   ...
> 
> }
> 
> ```
> 
> 
> 
> Looking forward to what @titusfortner thinks on that?

That sounds like a really nice idea, and it will definitely be better than my approach of just having urls that redirect to the error page

If @titusfortner agrees with this, I can update this PR tomorrow since there are only a couple of urls and then keep making small PRs for the rest of them

aguspe commented:
@p0deje I changed my implementation to match your suggestion and I fixed the rubocop issues, also I simplified the tests so I hope now this PR is on the right track :)

p0deje commented on rb/lib/selenium/webdriver/common/error.rb:
```diff
@@ -45,7 +52,7 @@ class WebDriverError < StandardError; end
       class NoSuchElementError < WebDriverError
         def initialize(msg = '')
-          super("#{msg}; #{SUPPORT_MSG} #{ERROR_URL}#no-such-element-exception")
+          super("#{msg}; #{SUPPORT_MSG} #{URLS[:NoSuchElementError]}")
```
Would it make sense to move this to WebDriverError constructor?

```ruby
if (url = URLS[self.class.name])
  msg = "#{msg}; #{SUPPORT_MSG} #{url}"
end
super(msg)
```

aguspe replied:
I was thinking the same, the reason why I didn't do it is that once we have all the pages added in the documentation, we will just refactor the WebDriverError constructor to be:

```Ruby
def initialize(msg = '')
  super("#{msg}; #{SUPPORT_MSG} #{URLS[self.class.name]}")
end
```
And that probably will be done by the last person updating the docs 

So Instead of having to add a conditional and remove it, I was thinking that is clearer for future maintainers to notice that there is a code duplication, but if you think is clearer with the conditional, I can just add it :) and a comment for the eventual removal

aguspe replied:
@p0deje I updated it now with a different style to avoid rubocop complaints and with a comment so in the future someone will remember to remove it

p0deje commented:
Thank you for the contribution!
